web/src/stores/workflow-store.ts

```diff
@@ -372,6 +381,36 @@ export const useWorkflowStore = create<WorkflowStoreState>((set, get) => ({
 			}),
 		);
 	},
+	hasMissingSecret: (nodeId: string, secretIdx?: number) => {
+		const missingSecrets = get().missingSecretForNodes;
+		if (missingSecrets) {
+			if (secretIdx !== undefined) {
+				return (
+					missingSecrets.has(nodeId) &&
+					missingSecrets.get(nodeId)!.has(secretIdx)
+				);
+			}
+			return missingSecrets.has(nodeId);
+		}
+		return false;
+	},
+	addMissingSecrets: (missingSecrets: Map<string, Set<number>>) =>
+		set(
+			produce((draft) => {
+				draft.missingSecretForNodes = missingSecrets;
+			}),
+		),
+	removeMissingSecretById: (nodeId: string, secretIdx: number) =>
+		set(
+			produce((draft) => {
+				if (draft.missingSecretForNodes.has(nodeId)) {
+					draft.missingSecretForNodes.get(nodeId).delete(secretIdx);
+					if (draft.missingSecretForNodes.get(nodeId).size === 0) {
+						draft.missingSecretForNodes.delete(nodeId);
+					}
+				}
```

Comment:
No need to perform lookup 3 times. I think you can just do:

```suggestion
				const secretPlaceholdersWithDeletedSecret = draft.missingSecretForNodes.get(nodeId);
				secretPlaceholdersWithDeletedSecret?.delete(secretIdx);
				if (secretPlaceholdersWithDeletedSecret?.size === 0) {
					draft.missingSecretForNodes.delete(nodeId);
				}
```